drivers: i2c: mcux: add support for target mode

mwerner96 opened the pull request:
Add i2c target support to the i2c_mcux driver.
Add the frdm_mcxc242 board to i2c_target_api tests.

mwerner96 committed:
drivers: i2c: mcux: add support for target mode

Add i2c target support to the i2c_mcux driver.
Add the frdm_mcxc242 board to i2c_target_api tests.

Signed-off-by: Maximilian Werner <maximilian.werner96@gmail.com>

Holt-Sun commented on drivers/i2c/i2c_mcux.c:
```diff
@@ -299,12 +307,170 @@ static int i2c_mcux_transfer_cb(const struct device *dev, struct i2c_msg *msgs,
+
+	I2C_MasterDeinit(base);
+
+	if (!target_config) {
```
The target_config->callbacks should be validated as well.

Holt-Sun commented on drivers/i2c/i2c_mcux.c:
```diff
@@ -299,12 +307,170 @@ static int i2c_mcux_transfer_cb(const struct device *dev, struct i2c_msg *msgs,
+}
+
+static int i2c_mcux_target_unregister(const struct device *dev,
+				      struct i2c_target_config *target_config)
```
target_config is unused.

Holt-Sun commented on tests/drivers/i2c/i2c_target_api/boards/frdm_mcxc242.overlay:
```diff
@@ -0,0 +1,51 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <nxp/mcx/MCXC142VFM-pinctrl.h>
```
Is it expected to load MCXC142VFM-pinctrl.h in frdm_mcxc242.overlay? Please verify.

mwerner96 replied:
That was a copy-paste error from one of my private boards. Thanks.

Holt-Sun commented on drivers/i2c/i2c_mcux.c:
```diff
@@ -299,12 +307,170 @@ static int i2c_mcux_transfer_cb(const struct device *dev, struct i2c_msg *msgs,
+	data->target_cfg = NULL;
+	data->target_attached = false;
+
+	I2C_SlaveDeinit(base);
```
Had better tear down slave mode first to avoid ISR using the wrong handle.

Holt-Sun commented:
1. Please fix the compliance issues.
2. Although the sonarqubecloud reported issue is not a blocker, but still suggest to fix because it is impact high which main affect maintenance.

mwerner96 committed:
drivers: i2c: mcux: refactor target event handler

The target event handler i2c_mcux_target_transfer_cb() had a too high
cognitive complexity. Create a function for each event type to reduce
the complexity score.

Signed-off-by: Maximilian Werner <maximilian.werner96@gmail.com>

butok commented on drivers/i2c/i2c_mcux.c:
```diff
@@ -299,12 +307,216 @@ static int i2c_mcux_transfer_cb(const struct device *dev, struct i2c_msg *msgs,
 
 #endif /* CONFIG_I2C_CALLBACK */
 
+#if CONFIG_I2C_TARGET
```
#if => #ifdef
